mlflow/utils/virtualenv.py

```diff
@@ -0,0 +1,223 @@
+import os
+import logging
+import shutil
+import uuid
+from pathlib import Path
+
+from mlflow.exceptions import MlflowException
+from mlflow.utils.process import _exec_cmd, _join_commands, _IS_UNIX
+from mlflow.utils.environment import (
+    PythonEnv,
+    _PYTHON_ENV_FILE_NAME,
+    _CONDA_ENV_FILE_NAME,
+    _get_mlflow_env_name,
+    _get_pip_install_mlflow,
+)
+from mlflow.utils.conda import _get_conda_dependencies
+
+
+_MLFLOW_ENV_ROOT_ENV_VAR = "MLFLOW_ENV_ROOT"
+
+
+_logger = logging.getLogger(__name__)
+
+
+def _get_mlflow_virtualenv_root():
+    """
+    Returns the root directory to store virtualenv environments created by MLflow.
+    """
+    return os.getenv(_MLFLOW_ENV_ROOT_ENV_VAR, str(Path.home().joinpath(".mlflow", "envs")))
+
+
+def _is_pyenv_available():
+    """
+    Returns True if pyenv is available, otherwise False.
+    """
+    return shutil.which("pyenv") is not None
+
+
+def _validate_pyenv_is_available():
+    """
+    Validates pyenv is available. If not, throws an `MlflowException` with a brief instruction on
+    how to install pyenv.
+    """
+    url = (
+        "https://github.com/pyenv/pyenv#installation"
+        if _IS_UNIX
+        else "https://github.com/pyenv-win/pyenv-win#installation"
+    )
+    if not _is_pyenv_available():
+        raise MlflowException(
+            f"Could not find the pyenv binary. See {url} for installation instructions."
+        )
+
+
+def _is_virtualenv_available():
+    """
+    Returns True if virtualenv is available, otherwise False.
+    """
+    return shutil.which("virtualenv") is not None
+
+
+def _validate_virtualenv_is_available():
+    """
+    Validates virtualenv is available. If not, throws an `MlflowException` with a brief instruction
+    on how to install virtualenv.
+    """
+    if not _is_virtualenv_available():
+        raise MlflowException(
+            "Could not find the virtualenv binary. Run `pip install virtualenv` to install "
+            "virtualenv."
+        )
+
+
+def _install_python(version):
+    """
+    Installs a specified version of python with pyenv and returns a path to the installed python
+    binary.
+
+    :param version: Python version to install.
+    :return: Path to the installed python binary.
+    """
+    _logger.info("Installing python %s", version)
+    # pyenv-win doesn't support `--skip-existing` but its behavior is enabled by default
+    # https://github.com/pyenv-win/pyenv-win/pull/314
+    pyenv_install_options = ("--skip-existing",) if _IS_UNIX else ()
+    _exec_cmd(["pyenv", "install", *pyenv_install_options, version], capture_output=False)
```

Comment:
This line expects `version` looks like `3.7.12`. Do we want to support `3.7`?

Reply:
Using `pyenv install --list`, we can list available python versions and find the latest one that matches `version`. For example, if `version` is `3.7`, select `3.7.12`.

Reply:
Great question! I think we should use the latest available micro version in that case. Great idea!

Reply:
Ok, I'll implement it.

Reply:
Implemeted.